internal/nexodus/keys.go

```diff
@@ -12,18 +14,112 @@ import (
 
 // default key pair file locations (windows needs work)
 const (
-	workdirPublicKeyFile  = "public.key"
-	workdirPrivateKeyFile = "private.key"
-	linuxPublicKeyFile    = "/etc/wireguard/public.key"
-	linuxPrivateKeyFile   = "/etc/wireguard/private.key"
-	darwinPublicKeyFile   = "/usr/local/etc/wireguard/public.key"
-	darwinPrivateKeyFile  = "/usr/local/etc/wireguard/private.key"
-	windowsPublicKeyFile  = "C:/nexd/public.key"
-	windowsPrivateKeyFile = "C:/nexd/private.key"
 	publicKeyPermissions  = 0644
 	privateKeyPermissions = 0600
 )
 
+// handleKeys will look for an existing key pair, if a pair is not found this method
+// will generate a new pair and write them to location on the disk depending on the OS
+func (ax *Nexodus) handleKeys() error {
+
+	pubKeyFile := filepath.Join(ax.stateDir, "public.key")
+	privKeyFile := filepath.Join(ax.stateDir, "private.key")
+
+	// We used to store the keys in a different location..
+	// migrate them to the state dir if needed.
+	err := ax.migrationCheck(pubKeyFile, privKeyFile)
+	if err != nil {
+		return err
+	}
+
+	publicKey := readKeyFile(ax.logger, pubKeyFile)
+	privateKey := readKeyFile(ax.logger, privKeyFile)
+	if publicKey != "" && privateKey != "" {
+		ax.wireguardPubKey = publicKey
+		ax.wireguardPvtKey = privateKey
+		ax.logger.Infof("Existing key pair found at [ %s ] and [ %s ]", pubKeyFile, privKeyFile)
+		return nil
+	}
+	ax.logger.Infof("No existing public/private key pair found, generating a new pair")
+	if err := ax.generateKeyPair(pubKeyFile, privKeyFile); err != nil {
+		return fmt.Errorf("Unable to locate or generate a key/pair: %w", err)
+	}
+	ax.logger.Debugf("New keys were written to [ %s ] and [ %s ]", pubKeyFile, privKeyFile)
+	return nil
+}
+
+// migrationCheck should not be needed after everyone has upgraded to the latest nexd.
+func (ax *Nexodus) migrationCheck(pubKeyFile string, privKeyFile string) error {
+
+	// skip if the new key files exist
+	if canReadFile(pubKeyFile) && canReadFile(privKeyFile) {
+		return nil
+	}
+
+	oldPubKeyFile := "public.key"
+	oldPrivKeyFile := "private.key"
+	if !ax.userspaceMode {
+		switch runtime.GOOS {
```

Comment:
@russellb so now with this change, even if you are in user space mode, we store the keys in the state dir.  I assume this difference is so that a non-root user could write the keys... I imagine that a user space mode user would typically configure the state dir right?

Reply:
I wouldn't expect a regular user to have write access to the dir where keys are stored. They shouldn't have read access, either (at least to the private key and any cached tokens).

For userspaceMode, I believe I had it just using the working directory? An improvement is probably a user-specific config dir. `~/.config/nexodus/` would probably work for linux and mac.

Reply:
Should we check to see if the process is running as root, then use the /var/lib dir otherwise use a ~/.config dir?

Reply:
I think we need two types of paths for two different types of files.

1. `stateDir` -- Variable State information (cached API tokens for now). This should default to `/var/lib/nexodus` for linux/mac and `C:/nexodus` for Windows.

2. `configDir` -- For nexodus configuration (wireguard keys for now). This should  remain in `/etc`. Creating a new `/etc/nexodus` is fine to have it isolated.

In both cases, if write access to the default paths is not available, fall back to `~/.nexodus` on linux/mac and just fail on Windows, since I don't know what's normal there.

Reply:
I see wireguard keys as being variable state information.  I don't really see why an admin would want to configure the wiregaurd keys.  Wouldn't he be happy with the ones generated by the system?   if he wants preserve identity after a move, then he needs to preserve the state dir.

Reply:
https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html

> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. **Users must never need to modify files in /var/lib to configure a package's operation, and the specific file hierarchy used to store the data must not be exposed to regular users.**

Reply:
Why does a user need to modify the keys?

Reply:
They represent the only unique per-device identity. If someone is already using wireguard, they may already have a repository of per-device keys. It's also possible that they may want to create a device ahead of time so they know upfront which IP is associated with each device.

We don't make this very easy to do in practice since neither the UI nor `nexctl` support creating a device.

I have a concrete use case for this. When running `nexd` in a Pod, I put the keys in a Secret, because I need `nexd` to come up as the same device, even after it gets restarted for some reason. Setting this up in advance is easier than trying to grab the keys from the first time `nexd` runs and stashing them.

I'm sure we can do better in the future with per-device API keys, but for now managing keys is the only identity control we have.